Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+        case .running:
+            self.state = .failed(error)
+            return .failRequest(error, .close, clearReadTimeoutTimer: false)
+        case .finished, .failed:
+            preconditionFailure("If the request is finished or failed, we expect the connection state machine to remove the request immediately from its state. Thus this state is unreachable.")
+        }
+    }
+
+    mutating func requestVerified(_ head: HTTPRequestHead) -> Action {
+        guard case .verifyingRequest = self.state else {
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+
+        guard self.isChannelWritable else {
+            self.state = .waitForChannelToBecomeWritable(head)
+            return .wait
+        }
+
+        return self.startSendingRequestHead(head)
+    }
+
+    mutating func requestStreamPartReceived(_ part: IOData) -> Action {
+        switch self.state {
+        case .initialized,
+             .waitForChannelToBecomeWritable,
+             .verifyingRequest,
+             .running(.endSent, _):
+            preconditionFailure("We must be in the request streaming phase, if we receive further body parts. Invalid state: \(self.state)")
+
+        case .running(.streaming(_, _, let producerState), .receivingBody(let head, _)) where head.status.code >= 300:
+            // If we have already received a response head with status >= 300, we won't send out any
+            // further request body bytes. Since the remote signaled with status >= 300, that it
+            // won't be interested. We expect that the producer has been informed to pause
+            // producing.
+            assert(producerState == .paused)
+            return .wait
+
+        case .running(.streaming(let expectedBodyLength, var sentBodyBytes, let producerState), let responseState):
+            // We don't check the producer state here:
+            //
+            // No matter if the `producerState` is either `.producing` or `.paused` any bytes we
+            // receive shall be forwarded to the Channel right away. As long as we have not received
+            // a response with status >= 300.
+            //
+            // More streamed data is accepted, even though the producer may have been asked to
+            // pause. The reason for this is as follows: There might be thread synchronization
+            // situations in which the producer might not have received the plea to pause yet.
+
+            if let expected = expectedBodyLength {
+                if sentBodyBytes + part.readableBytes > expected {
+                    let error = HTTPClientError.bodyLengthMismatch
+
+                    var clearReadTimeoutTimer = false
+                    if case .receivingBody = responseState, self.idleReadTimeout != nil {
+                        clearReadTimeoutTimer = true
+                    }
+
+                    return .failRequest(error, .close, clearReadTimeoutTimer: clearReadTimeoutTimer)
```

Comment:
Is there any reason this state machine has to keep track of whether we need to clear the read timeout timer? Presumably someone has set the timer, and they know if they have one or not, so should unconditionally clear it.